### PR TITLE
Add Session-based RecSys and Sequential RecSys, and AAAI'19_SR-GNN Paper

### PR DESCRIPTION
**Pull request by kingmbc**
## PR
- Add Session-based RecSys and Sequential RecSys
- Add 19AAAI_SR-GNN Paper

### CODE REVIEW

File: SessionBasedRec/19AAAI_SR-GNN/official/datasets/preprocess.py [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+    for s, date in tra_sess:
+        seq = sess_clicks[s]
+        outseq = []
+        for i in seq:
+            if i in item_dict:
+                outseq += [item_dict[i]]
+            else:
+                outseq += [item_ctr]
+                item_dict[i] = item_ctr
+                item_ctr += 1
```
Comment (kingmbc) [inline SessionBasedRec/19AAAI_SR-GNN/official/datasets/preprocess.py:148]: item_id가 1부터 시작함.

File: SessionBasedRec/19AAAI_SR-GNN/official/pytorch_code/model.py [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+import datetime
+import math
+import numpy as np
+import torch
+from torch import nn
+from torch.nn import Module, Parameter
+import torch.nn.functional as F
+
+
+class GNN(Module):
```
Comment (kingmbc) [inline SessionBasedRec/19AAAI_SR-GNN/official/pytorch_code/model.py:18]: GGNN 구현

File: SessionBasedRec/19AAAI_SR-GNN/official/pytorch_code/model.py [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+    def compute_scores(self, hidden, mask):
+        ht = hidden[torch.arange(mask.shape[0]).long(), torch.sum(mask, 1) - 1]  # batch_size x latent_size
+        q1 = self.linear_one(ht).view(ht.shape[0], 1, ht.shape[1])  # batch_size x 1 x latent_size
+        q2 = self.linear_two(hidden)  # batch_size x seq_length x latent_size
+        alpha = self.linear_three(torch.sigmoid(q1 + q2))
+        a = torch.sum(alpha * hidden * mask.view(mask.shape[0], -1, 1).float(), 1)
+        if not self.nonhybrid:
+            a = self.linear_transform(torch.cat([a, ht], 1))
+        b = self.embedding.weight[1:]  # n_nodes x latent_size
```
Comment (kingmbc) [inline SessionBasedRec/19AAAI_SR-GNN/official/pytorch_code/model.py:87]: item index가 1부터 시작하고, 0는 PAD로 씀.
따라서 0은 무시하고 계산하기 위함

File: SessionBasedRec/19AAAI_SR-GNN/official/pytorch_code/model.py [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+    model.scheduler.step()
+    print('start training: ', datetime.datetime.now())
+    model.train()
+    total_loss = 0.0
+    slices = train_data.generate_batch(model.batch_size)
+    for i, j in zip(slices, np.arange(len(slices))):
+        model.optimizer.zero_grad()
+        targets, scores = forward(model, i, train_data)
+        targets = trans_to_cuda(torch.Tensor(targets).long())
+        loss = model.loss_function(scores, targets - 1)
```
Comment (kingmbc) [inline SessionBasedRec/19AAAI_SR-GNN/official/pytorch_code/model.py:133]: 여기서도 targets - 1을 하는게 item index가 1부터 시작하기 때문이고,
실제 label을 제대로 값을 맞춰주기 위함.

File: SessionBasedRec/19AAAI_SR-GNN/official/pytorch_code/utils.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+            for j, i in graph.in_edges(i):
+                graph.add_edge(j, i, weight=graph.get_edge_data(j, i)['weight'] / sum)
+    return graph
+
+
+def data_masks(all_usr_pois, item_tail):
+    us_lens = [len(upois) for upois in all_usr_pois]
+    len_max = max(us_lens)
+    us_pois = [upois + item_tail * (len_max - le) for upois, le in zip(all_usr_pois, us_lens)]
+    us_msks = [[1] * le + [0] * (len_max - le) for le in us_lens]
```
Comment (kingmbc) [inline SessionBasedRec/19AAAI_SR-GNN/official/pytorch_code/utils.py:36]: [0]가 PAD 임

File: SessionBasedRec/19AAAI_SR-GNN/official/pytorch_code/utils.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+                u = np.where(node == u_input[i])[0][0]
+                v = np.where(node == u_input[i + 1])[0][0]
+                u_A[u][v] = 1
+            u_sum_in = np.sum(u_A, 0)
+            u_sum_in[np.where(u_sum_in == 0)] = 1
+            u_A_in = np.divide(u_A, u_sum_in)
+            u_sum_out = np.sum(u_A, 1)
+            u_sum_out[np.where(u_sum_out == 0)] = 1
+            u_A_out = np.divide(u_A.transpose(), u_sum_out)
+            u_A = np.concatenate([u_A_in, u_A_out]).transpose()
```
Comment (kingmbc) [inline SessionBasedRec/19AAAI_SR-GNN/official/pytorch_code/utils.py:102]: GGNN에 들어가기 전에 2가지의 Adj. matrix를 구하고 Degree로 Normalize해줌

image

File: SessionBasedRec/19AAAI_SR-GNN/official/pytorch_code/utils.py [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+    train_set_x = [train_set_x[s] for s in sidx[:n_train]]
+    train_set_y = [train_set_y[s] for s in sidx[:n_train]]
+
+    return (train_set_x, train_set_y), (valid_set_x, valid_set_y)
+
+
+class Data():
+    def __init__(self, data, shuffle=False, graph=None):
+        inputs = data[0]
+        inputs, mask, len_max = data_masks(inputs, [0])
```
Comment (kingmbc) [inline SessionBasedRec/19AAAI_SR-GNN/official/pytorch_code/utils.py:57]: input에 session마다 길이가 다를수 있는데, truncate하지 않고, 전체를 다 씀

**Review by kingmbc — COMMENTED** [review]
끝
